datafusion/optimizer/src/type_coercion.rs

```diff
@@ -98,6 +101,81 @@ impl ExprRewriter for TypeCoercionRewriter<'_> {
 
     fn mutate(&mut self, expr: Expr) -> Result<Expr> {
         match expr {
+            Expr::IsTrue(expr) => {
+                let expr = is_true(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::IsNotTrue(expr) => {
+                let expr = is_not_true(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::IsFalse(expr) => {
+                let expr = is_false(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::IsNotFalse(expr) => {
+                let expr =
+                    is_not_false(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::Like {
+                negated,
+                expr,
+                pattern,
+                escape_char,
+            } => {
+                let left_type = expr.get_type(&self.schema)?;
+                let right_type = pattern.get_type(&self.schema)?;
+                let coerced_type =
+                    coerce_types(&left_type, &Operator::Like, &right_type)?;
+                let expr = Box::new(expr.cast_to(&coerced_type, &self.schema)?);
+                let pattern = Box::new(pattern.cast_to(&coerced_type, &self.schema)?);
+                let expr = Expr::Like {
+                    negated,
+                    expr,
+                    pattern,
+                    escape_char,
+                };
+                expr.rewrite(&mut self.const_evaluator)
```

Comment:
🤔  maybe in the long term figuring out how to avoid having to remember to call the const evaluator each struct would be good -- as it too does a expr tree walk

So this algorithm is at least `O(n^2)` in the number of items in the expr tree. 

I don't think it is important to do in this PR, but I think it is worth keeping in mind as we proceed or if we see the planner being a bit slow

Reply:
> maybe in the long term figuring out how to avoid having to remember to call the const evaluator each struct would be good -- as it too does a expr tree walk

agree with that

Reply:
I think the SQL level with analyse/expr/optimizer should be refactor.
We are missing some formal logic like spark SQL framework or other SQL framework.
This makes it very difficult for us to do some features or optimization.

Reply:
I agree the current framework for expr rewriting / analysis is reaching the limits of its existing design. Refactoring the code so it is easier to work with sounds like a good idea to me. 

I am open to suggestions -- I think the challenge in any such refactors is not only coming up with a good new design but also figuring out how to incrementally get the existing code into that new structure